// style-helper/dist/index.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var free_style_1 = require("free-style");
var index_1 = require("./index");
describe('style helper', function () {
    it('should quote a string', function () {
        expect(index_1.quote('testing')).toEqual('"testing"');
        expect(index_1.quote('"')).toEqual('"\\""');
    });
    it('should handle urls', function () {
        expect(index_1.url('/path.png')).toEqual('url("/path.png")');
    });
    it('should turn styles into an object', function () {
        expect(index_1.objectify('padding', 10)).toEqual({
            padding: 10
        });
    });
    it('should merge css object together', function () {
        expect(index_1.merge({
            margin: 5,
            padding: 10
        }, {
            '@media print': {
                margin: 0,
                padding: 0
            }
        }, {
            '@media print': {
                padding: 5
            }
        })).toEqual({
            margin: 5,
            padding: 10,
            '@media print': {
                margin: 0,
                padding: 5
            }
        });
    });
    it('should register style sheets', function () {
        var Style = free_style_1.create();
        var styles = index_1.registerStyleSheet(Style, {
            button: {
                color: 'red'
            }
        }, {
            css: {
                html: {
                    margin: 0
                }
            }
        });
        expect(Object.keys(styles)).toEqual(['button']);
        expect(Style.getStyles()).toEqual("." + styles.button + "{color:red}html{margin:0}");
    });
});
